File: Tree/94BinaryTreeInorderTraversal.cpp
/**
 In-order Traversal
 In-order traversal is to traverse the left subtree first. Then visit the root. Finally, traverse the right subtree.
 
 Solution:
 Create a vector to store each node accessed. Use a helper function to traverse nodes recursively. Return NULL when current node is NULL.
 Step 1: Go to left node of current node.
 Step 2: When there is no more left node, store the current node's value.
 Step 3: Go to right node of current node.
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> rev;
        helper(root, rev);
        return rev;
    }
    void helper(TreeNode* root, vector<int> &rev){
        if(root==NULL) return;
        else{
            
            helper(root->left, rev);
            rev.push_back(root->val);
            helper(root->right, rev);
        }
    }
};

File: Tree/94binaryTreeInorderTraversal.py
#key: BinaryTree
#Algorithm: recursion

#like the problem 144, the ans.append should occur behind the root.left
#

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ans = []
        
        def check(root):
            if not root:
                return
            else:
                check(root.left)
                ans.append(root.val)
                check(root.right)
        
        check(root)
        return ans